File: src/nbla/function/generic/fused_batch_normalization.cpp
// Copyright (c) 2017 Sony Corporation. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <nbla/array.hpp>
#include <nbla/function/fused_batch_normalization.hpp>
#include <nbla/variable.hpp>

#include <algorithm>
#include <cmath>
#include <limits>
#include <numeric>

namespace nbla {

NBLA_REGISTER_FUNCTION_SOURCE(FusedBatchNormalization, const vector<int> &,
                              float, float, bool, const string &);

namespace fused_batch_normalization {
// These functions are special cases for the fused batch normalization
template <typename T>
void relu_backward(int size, T *dx, const T *dy, const T *y) {
  for (int i = 0; i < size; i++) {
    if (y[i] > 0)
      dx[i] = dy[i];
    else
      dx[i] = T(0);
  }
}

template <typename T>
void add2_backward(int size, T *dx1, const T *dx, bool accum) {
  for (int i = 0; i < size; i++) {
    dx1[i] = accum ? dx1[i] + dx[i] : dx[i];
  }
}
}

template <class T>
void FusedBatchNormalization<T>::setup_impl(const Variables &inputs,
                                            const Variables &outputs) {
  NBLA_CHECK(nonlinearity_ == "relu", error_code::not_implemented,
             "Currently \"relu\" is only supported as a nonlinearity.");
  Variables inputs_bn(inputs.begin(), inputs.begin() + 5);
  bn_ = create_BatchNormalization(this->ctx_, axes_, decay_rate_, eps_,
                                  batch_stat_, false /* no_scale */,
                                  false /* no_bias */);
  bn_->setup(inputs_bn, outputs);
}

template <class T>
void FusedBatchNormalization<T>::forward_impl(const Variables &inputs,
                                              const Variables &outputs) {

  NBLA_CHECK(bn_, error_code::value, "setup is not called.");
  // Naive non-fused implementation by layer composition.
  // 1. Perform BatchNormalization
  Variables inputs_bn(inputs.begin(), inputs.begin() + 5);
  bn_->forward(inputs_bn, outputs);

  // 2. Perform Add2
  // NOTE: Output buffer are re-used by inplacing.
  if (inputs.size() == 6) {
    auto add2 = create_Add2(this->ctx_, true);
    add2->setup(Variables{outputs[0], inputs[5]}, Variables{outputs[0]});
    add2->forward(Variables{outputs[0], inputs[5]}, Variables{outputs[0]});
  }

  // 3. Perform ReLU
  // NOTE: Output buffer are re-used by inplacing.
  auto relu = create_ReLU(this->ctx_, true);
  relu->setup(Variables{outputs[0]}, Variables{outputs[0]});
  relu->forward(Variables{outputs[0]}, Variables{outputs[0]});
}

template <class T>
void FusedBatchNormalization<T>::backward_impl(
    const Variables &inputs, const Variables &outputs,
    const vector<bool> &propagate_down, const vector<bool> &accum) {
  NBLA_CHECK(bn_, error_code::value, "setup is not called.");

  // Naive non-fused implementation by layer composition.
  // 1. Perform ReLU backward
  // NOTE: Output buffer are re-used by inplacing.
  bool prop_down_add2 = (inputs.size() == 6 && propagate_down[5]);
  bool prop_down_bn =
      std::accumulate(propagate_down.begin(), propagate_down.begin() + 3, false,
                      std::logical_or<bool>());
  auto y = outputs[0]->get_data_pointer<T>(this->ctx_);
  auto dx = outputs[0]->cast_grad_and_get_pointer<T>(this->ctx_);
  auto dy = outputs[0]->get_grad_pointer<T>(this->ctx_);
  auto size = outputs[0]->size();
  if (prop_down_add2 || prop_down_bn) {
    fused_batch_normalization::relu_backward(size, dx, dy, y);
  }

  // 2. Perform Add2 backward
  // NOTE: Output buffer for the first operand of the addition are re-used by
  // inplacing,
  // nothing done for it.
  if (prop_down_add2) {
    auto dx1 = inputs[5]->cast_grad_and_get_pointer<T>(this->ctx_);
    fused_batch_normalization::add2_backward(size, dx1, dx, accum[5]);
  }
  // 3. Perform BN backward
  Variables inputs_bn(inputs.begin(), inputs.begin() + 5);
  vector<bool> prop_down_bn_inputs(propagate_down.begin(),
                                   propagate_down.begin() + 5);
  vector<bool> accum_bn_inputs(accum.begin(), accum.begin() + 5);
  bn_->backward(inputs_bn, outputs, prop_down_bn_inputs, accum_bn_inputs);
}
} // namespace nbla

File: python/src/nnabla/experimental/graph_converters/fused_batch_normalization.py
# Copyright 2020,2021 Sony Corporation.
# Copyright 2021 Sony Group Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import nnabla.parametric_functions as PF

from .graph_converter import FunctionModifier


class FusedBatchNormalizationModifier(FunctionModifier):
    """
    Block `BatchNormalization -> Add2 -> Non-Linear` pass is fused into one `FusedBatchNormalization`.

    If there is a block `BatchNormalization -> Add2 -> Non-Linear` pass,
    remove all the block functions and replace the whole block to `FusedBatchNormalization`.

    Examples:

    .. code-block:: python

       pred = Model(...)

       import nnabla.experimental.graph_converters as GC

       modifiers = [GC.FusedBatchNormalizationModifier()]
       gc = GC.GraphConverter(modifiers)
       pred = gc.convert(pred)
    """

    def __init__(self):
        super(FusedBatchNormalizationModifier, self).__init__()
        self._name = ''
        self._block = False
        self._bn_args = None
        self._add2_input1 = None
        self._cnt = 1
        self._fct_set = {
            'ReLU': 'relu'
        }

    def modify(self, f, inputs):
        outputs = f.outputs[0]

        # Not end
        if len(outputs.function_references) == 0:
            return

        # Check fused bn block start
        if not self._block and self._is_fused_bn_block(f, inputs):
            self._block = True

        # Remove BatchNormalization
        if self._block and f.info.type_name == 'BatchNormalization':
            self._bn_args = f.info.args
            self._name = self.get_parameter_scope(inputs[0])
            return inputs[0]

        # Remove Add2
        if self._block and f.info.type_name == 'Add2':
            self._add2_input1 = inputs[1]
            return inputs[0]

        # Remove non linear function then connect fused bn
        if self._block and f.info.type_name in self._fct_set:
            f_non_linear = self._fct_set[f.info.type_name]
            h = PF.fused_batch_normalization(
                inputs[0], self._add2_input1,
                nonlinearity=f_non_linear, **self._bn_args,
                name='fused{}-{}'.format(self._name, self._cnt))
            self._cnt += 1
            self._block = False
            return h

    def _is_fused_bn_block(self, f, inputs):
        outputs = f.outputs[0]

        # Function is BN whose next function is Add2,
        # function after Add2 is not non-linear
        next_func = outputs.function_references[0]

        if len(next_func.outputs[0].function_references) == 0:
            return False

        nnext_func = next_func.outputs[0].function_references[0]

        if f.info.type_name == 'BatchNormalization' \
           and next_func.info.type_name == 'Add2'   \
           and nnext_func.info.type_name in self._fct_set:
            return True

        return False

    def __finish__(self):
        self._name = ''
        self._block = False
        self._bn_args = None
        self._add2_input1 = None
        self._cnt = 1
